### PR TITLE
Extract validation

### PR DESCRIPTION
**Pull request by pepyakin**
Extract validation into separate module. 
After this change, you can validate module without instantiating it. 
For the moment, old validation in instantiation still in place, this will be fixed in follow-up PRs.

Notable changes:

- common code moved into `common` module (i.e. `StackWithLimit`),
- separate validation tests added under `validation::tests`,
- `interpreter::validator` is copied into `validation::func`,
- Require leading `end` for init exprs,

[Validation spec](https://webassembly.github.io/spec/valid/index.html) for reviewer convenience

### CODE REVIEW

File: src/interpreter/runner.rs [hunk 1]
```diff
@@ -541,26 +551,39 @@ impl Interpreter {
 	}
 
 	fn run_current_memory<'a>(context: &mut FunctionContext) -> Result<InstructionOutcome<'a>, Error> {
-		context.module()
+		context
+			.module()
 			.memory(ItemIndex::IndexSpace(DEFAULT_MEMORY_INDEX))
 			.map(|m| m.size())
-			.and_then(|s| context.value_stack_mut().push(RuntimeValue::I32(s as i32)))
+			.and_then(|s| {
```
Comment (NikVolf) [inline src/interpreter/runner.rs]: why add braces here?

File: src/interpreter/runner.rs [hunk 2]
```diff
@@ -541,26 +551,39 @@ impl Interpreter {
...
 
 	fn run_grow_memory<'a>(context: &mut FunctionContext) -> Result<InstructionOutcome<'a>, Error> {
 		let pages: u32 = context.value_stack_mut().pop_as()?;
-		context.module()
+		context
+			.module()
 			.memory(ItemIndex::IndexSpace(DEFAULT_MEMORY_INDEX))
 			.and_then(|m| m.grow(pages))
-			.and_then(|m| context.value_stack_mut().push(RuntimeValue::I32(m as i32)))
+			.and_then(|m| {
```
Comment (NikVolf) [inline src/interpreter/runner.rs]: ditto

File: src/validation/func.rs [hunk 3]
```diff
@@ -0,0 +1,773 @@
...
+			context.position += 1;
+			if context.position >= body_len {
+				return Ok(());
+			}
+		}
+	}
+
+	fn validate_instruction(context: &mut FunctionValidationContext, opcode: &Opcode) -> Result<InstructionOutcome, Error> {
+		debug!(target: "validator", "validating {:?}", opcode);
+		match opcode {
```
Comment (NikVolf) [inline src/validation/func.rs]: if this already in the diff and pr is about validation won't you mind changing to `match *opcode { ... }`

File: src/validation/mod.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+	let mut memories = Vec::new();
+	let mut globals = Vec::new();
+
+	for import_entry in module
+		.import_section()
+		.map(|i| i.entries())
+		.unwrap_or_default()
+	{
+		match import_entry.external() {
+			&External::Function(idx) => func_type_indexes.push(idx),
```
Comment (NikVolf) [inline src/validation/mod.rs]: ditto `match *import_entry.external() { .. }`

File: src/validation/tests.rs [hunk 5]
```diff
@@ -0,0 +1,299 @@
+use super::validate_module;
+use builder::module;
+use elements::{External, GlobalEntry, GlobalType, ImportEntry, InitExpr, MemoryType,
+               Opcode, Opcodes, TableType, ValueType, BlockType};
```
Comment (NikVolf) [inline src/validation/tests.rs]: should be just single tab as itendation
or
```rust
use elements::{
     ...,
     ...,
}
```

File: src/validation/mod.rs [hunk 6]
```diff
@@ -0,0 +1,319 @@
+#![allow(missing_docs)]
+
+use std::fmt;
+use elements::{BlockType, External, GlobalEntry, GlobalType, Internal, MemoryType,
```
Comment (NikVolf) [inline src/validation/mod.rs]: same about single itendation / dedicated lines

File: src/validation/func.rs (hunk 3 above)
Comment (pepyakin) [reply src/validation/func.rs]: yeah sure

File: src/validation/context.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			None => {
+				return Err(Error(format!("Table at index {} doesn't exists", idx)));
+			}
+		};
+
+		Ok(table)
+	}
+
+	pub fn require_function(&self, idx: u32) -> Result<(&[ValueType], BlockType), Error> {
+		let ty_idx = match self.func_type_indexes().get(idx as usize) {
```
Comment (NikVolf) [inline src/validation/context.rs]: *self.func_type_indexes().get(..).ok_or()?

File: src/validation/context.rs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			.unwrap_or(BlockType::NoResult);
+		Ok((params, return_ty))
+	}
+
+	pub fn require_global(
+		&self,
+		idx: u32,
+		mutability: Option<bool>,
+	) -> Result<&GlobalType, Error> {
+		let global = match self.globals().get(idx as usize) {
```
Comment (NikVolf) [inline src/validation/context.rs]:
```
let global = self.globals().get(idx as usize)
    .ok_or(Error(format!("Global at index {} doesn't exists", idx)))?;
```

File: src/validation/func.rs [hunk 9]
```diff
@@ -0,0 +1,773 @@
...
+
+		loop {
+			let opcode = &body[context.position];
+			match Validator::validate_instruction(context, opcode)? {
+				InstructionOutcome::ValidateNextInstruction => (),
+				InstructionOutcome::Unreachable => context.unreachable()?,
+			}
+
+			context.position += 1;
+			if context.position >= body_len {
```
Comment (NikVolf) [inline src/validation/func.rs]: `if context.position == body_len`

File: src/validation/func.rs [hunk 10]
```diff
@@ -0,0 +1,773 @@
...
+
+			context.position += 1;
+			if context.position == body_len {
+				return Ok(());
+			}
+		}
+	}
+
+	fn validate_instruction(context: &mut FunctionValidationContext, opcode: &Opcode) -> Result<InstructionOutcome, Error> {
+		debug!(target: "validator", "validating {:?}", opcode);
```
Comment (NikVolf) [inline src/validation/func.rs]: use Opcode::*; ? :)

**Comment by NikVolf** [comment]
Great job!

**Pull request merged**